templates/usr/local/sbin/simp_generate_types.epp

```diff
@@ -3,132 +3,210 @@
+CMD_NAME = File.basename($0)
 
-puppet_cmd = '/opt/puppetlabs/bin/puppet'
-
-puppet_group = %x(#{puppet_cmd} config print --section=master group).strip
-
-environment_paths = Dir.glob(
-  %x{#{puppet_cmd} config print --section=master environmentpath}.strip.split(':').map{|x| x + '/*'}
-)
+LOG = Syslog::Logger.new(File.basename(__FILE__))
 
 options = OpenStruct.new
 options.quiet = false
 options.path = nil
-options.delay = 0
+options.generate = false
+options.delay = 10
 options.syslog = false
-options.target_environments = []
 
 OptionParser.new do |opts|
   opts.on('-p', '--path TGT_PATH') do |path|
     options.path = path
-
-    if path == 'all'
-      options.target_environments = environment_paths.compact
-    else
-      path << '/' if (File.directory?(path) and path[-1] != '/')
-      options.target_environments = Array(environment_paths.find do |env|
-        path.start_with?("#{env}/")
-      end).compact
-    end
-
-    # Convert each discovered environment into a hash noting the path and environment name
-    options.target_environments.map!{|e| { :path => e, :env_name => File.basename(e) }}
   end
-  opts.on('-l', '--logfile LOGFILE') do |file|
+  opts.on('-g', '--generate', 'Read TGT_PATH and process all items') do
+    options.generate = true
+  end
+  opts.on('-l', '--logfile LOGFILE', 'Output to this logfile') do |file|
     options.logfile = file
   end
-  opts.on('-d', '--delay SECONDS') do |delay|
+  opts.on('-d', '--delay SECONDS', 'Default: 10 - Wait a random number of seconds up to this limit before starting the reader') do |delay|
     options.delay = delay.to_i
   end
-  opts.on('-s', '--syslog') do
+  opts.on('-s', '--syslog', 'Write to syslog') do
     options.syslog = true
   end
-  opts.on('-q', '--quiet') do
+  opts.on('-q', '--quiet', 'No terminal output') do
     options.quiet = true
   end
+  opts.on('-m', '--message MESSAGE', 'The message to be written to TGT_PATH from -p') do |msg|
+    options.message = msg.strip
+  end
   opts.on('-h', '--help', 'This help message') do
     puts opts
     exit
   end
 
   options.help = opts.help
-
 end.parse!
 
-fh = File.open(options.logfile, 'w') if options.logfile
-
-unless options.path
-  msg = "Error: You must pass a path"
+def lock(lockfile)
+  File.open(lockfile, 'w'){|fh| fh.puts(Process.pid)}
+end
 
-  $stderr.puts msg
-  $stderr.puts options.help
+def unlock(lockfile)
+  FileUtils.rm_f(lockfile) if File.exist?(lockfile)
+end
 
-  log.error(msg) if options.syslog
+def wait_and_lock(lockfile, timeout=30)
+  if File.exist?(lockfile)
+    begin
+      Timeout::timeout(timeout) do
+        while(File.exist?(lockfile)) do
+          sleep(2)
+        end
+      end
+    rescue Timeout::TimeoutError
+      LOG.error("Timed out waiting for '#{lockfile}' to release. Please remove the file manually if no '#{File.basename(CMD_NAME)}' processes are running.")
+
+      exit 1
+    end
+  end
 
-  fh.puts(msg) if fh
+  at_exit { unlock(lockfile) }
```

Comment:
This will remove the lock file, even if the process timed out waiting for another process to release the lock.  So, you can end up with a state in which the hung or long-running process is still running but no longer locked.  Furthermore, the log message on line 70 is no longer applicable.
So, IMHO you have a couple of options at_exit,
* Remove the file if your PID is in the file
* Kill the PID if it isn't you and then remove the file